hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/test/java/org/apache/hadoop/yarn/client/TestNoHaRMFailoverProxyProvider.java

```diff
@@ -0,0 +1,272 @@
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Unit tests for {@link DefaultNoHARMFailoverProxyProvider} and
+ * {@link AutoRefreshNoHARMFailoverProxyProvider}.
+ */
+public class TestNoHaRMFailoverProxyProvider {
+
+  // Default port of yarn RM
+  private static final int RM1_PORT = 8032;
+  private static final int RM2_PORT = 8031;
+
+  private static final int NUMNODEMANAGERS = 1;
+  private Configuration conf;
+
+  @Before
+  public void setUp() throws IOException, YarnException {
+    conf = new YarnConfiguration();
+  }
+
+  /**
+   * Tests the proxy generated by {@link DefaultNoHAFailoverProxyProvider}
+   * will connect to RM.
+   */
+  @Test
+  public void testRestartedRM() throws Exception {
+    MiniYARNCluster cluster =
+        new MiniYARNCluster("testRestartedRMNegative", NUMNODEMANAGERS, 1, 1);
+    YarnClient rmClient = YarnClient.createYarnClient();
+    try {
+      cluster.init(conf);
+      cluster.start();
+      final Configuration yarnConf = cluster.getConfig();
+      rmClient = YarnClient.createYarnClient();
+      rmClient.init(yarnConf);
+      rmClient.start();
+      List <NodeReport> nodeReports = rmClient.getNodeReports();
+      assertEquals(
+          "The proxy didn't get expected number of node reports",
+          NUMNODEMANAGERS, nodeReports.size());
+    } finally {
+      if (rmClient != null) {
+        rmClient.stop();
+      }
+      cluster.stop();
+    }
+  }
+
+  /**
+   * Tests the proxy generated by
+   * {@link AutoRefreshNoHARMFailoverProxyProvider} will connect to RM.
+   */
+  @Test
+  public void testConnectingToRM() throws Exception {
+    conf.setClass(YarnConfiguration.CLIENT_FAILOVER_NO_HA_PROXY_PROVIDER,
+        AutoRefreshNoHARMFailoverProxyProvider.class,
+        RMFailoverProxyProvider.class);
+    MiniYARNCluster cluster =
+        new MiniYARNCluster("testRestartedRMNegative", NUMNODEMANAGERS, 1, 1);
+    YarnClient rmClient = null;
+    try {
+      cluster.init(conf);
+      cluster.start();
+      final Configuration yarnConf = cluster.getConfig();
+      rmClient = YarnClient.createYarnClient();
+      rmClient.init(yarnConf);
+      rmClient.start();
+      List <NodeReport> nodeReports = rmClient.getNodeReports();
+      assertEquals(
+          "The proxy didn't get expected number of node reports",
+          NUMNODEMANAGERS, nodeReports.size());
+    } finally {
+      if (rmClient != null) {
+        rmClient.stop();
+      }
+      cluster.stop();
+    }
+  }
+
+  /**
+   * Test that the {@link DefaultNoHARMFailoverProxyProvider}
+   * will generate different proxies after RM IP changed
+   * and {@link DefaultNoHARMFailoverProxyProvider#performFailover(Object)}
+   * get called.
+   */
+  @Test
+  public void testDefaultFPPGetOneProxy() throws Exception {
+    // Create a proxy and mock a RMProxy
+    Proxy mockProxy1 = new TestProxy((proxy, method, args) -> null);
```

Comment:
It looks likes Yetus is complaining about this:
error: Proxy(InvocationHandler) has protected